File: languages/fsharp/exercises/concept/lists/ListsTests.fs
module ListsTests

open FsUnit.Xunit
open Xunit

open Lists

[<Fact>]
let ``New list``() =
    let expected: string list = []
    newList |> should equal expected

[<Fact(Skip = "Remove to run test")>]
let ``Existing list``() = existingList |> should equal [ "F#"; "Clojure"; "Haskell" ]

[<Fact(Skip = "Remove to run test")>]
let ``Add language to new list``() = addLanguage "Scala" newList |> should equal [ "Scala" ]

[<Fact(Skip = "Remove to run test")>]
let ``Add language to existing list``() =
    addLanguage "Common Lisp" existingList |> should equal [ "Common Lisp"; "F#"; "Clojure"; "Haskell" ]

[<Fact(Skip = "Remove to run test")>]
let ``Add language to custom list``() = addLanguage "Racket" [ "Scheme" ] |> should equal [ "Racket"; "Scheme" ]

[<Fact(Skip = "Remove to run test")>]
let ``Last added language of existing list``() = lastAddedLanguage existingList |> should equal "F#"

[<Fact(Skip = "Remove to run test")>]
let ``Last added language of custom list``() = lastAddedLanguage [ "Crystal"; "Nim" ] |> should equal "Crystal"

[<Fact(Skip = "Remove to run test")>]
let ``Count languages on new list``() = countLanguages newList |> should equal 0

[<Fact(Skip = "Remove to run test")>]
let ``Count languages on existing list``() = countLanguages existingList |> should equal 3

[<Fact(Skip = "Remove to run test")>]
let ``Count languages on custom list``() = countLanguages [ "Python"; "JavaScript" ] |> should equal 2

[<Fact(Skip = "Remove to run test")>]
let ``Reverse order of new list``() =
    let expected: string list = []
    reverseList newList |> should equal expected

[<Fact(Skip = "Remove to run test")>]
let ``Reverse order of existing list``() = reverseList existingList |> should equal [ "Haskell"; "Clojure"; "F#" ]

[<Fact(Skip = "Remove to run test")>]
let ``Reverse order of custom list``() =
    reverseList [ "Kotlin"; "Java"; "Scala"; "Clojure" ] |> should equal [ "Clojure"; "Scala"; "Java"; "Kotlin" ]

File: languages/csharp/exercises/concept/lists/ListsTests.cs
using Xunit;

public class ListsTests
{
    [Fact]
    public void NewList()
    {
        Assert.Empty(Languages.NewList());
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void ExistingList()
    {
        Assert.Equal(new string[] {"C#", "Clojure", "Elm"}, Languages.GetExistingLanguages());
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void AddLanguage()
    {
        Assert.Equal(new string[] {"C#", "Clojure", "Elm", "Bash"},
            Languages.AddLanguage(Languages.GetExistingLanguages(), "Bash"));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void ReverseList()
    {
        Assert.Equal(new string[] {"Elm", "Clojure", "C#"},
            Languages.ReverseList(Languages.GetExistingLanguages()));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void ContainsStar_yes()
    {
        Assert.True(Languages.ContainsStar(Languages.GetExistingLanguages()));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void ContainsStar_too_many()
    {
        var languages = Languages.GetExistingLanguages();
        languages.Insert(0, "VBA");
        Assert.False(Languages.ContainsStar(languages));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void ContainsStar_empty()
    {
        var languages = Languages.NewList();
        Assert.False(Languages.ContainsStar(languages));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void ContainsStar_single_star()
    {
        var languages = Languages.GetExistingLanguages();
        languages.RemoveAt(2);
        languages.RemoveAt(1);
        Assert.True(Languages.ContainsStar(languages));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void RemoveLanguage_yes()
    {
        var languages = Languages.GetExistingLanguages();
        Assert.Equal(new string[] {"C#", "Elm"}, Languages.RemoveLanguage(languages, "Clojure"));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void RemoveLanguage_no()
    {
        var languages = Languages.GetExistingLanguages();
        Assert.Equal(new string[] {"C#", "Clojure", "Elm"}, Languages.RemoveLanguage(languages, "English"));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void EnsureUnique_yes()
    {
        var languages = Languages.GetExistingLanguages();
        Assert.True(Languages.EnsureUnique(languages));
    }

    [Fact/*(Skip = "Remove this Skip property to run this test")*/]
    public void EnsureUnique_no()
    {
        var languages = Languages.GetExistingLanguages();
        languages.Add("C#");
        Assert.False(Languages.EnsureUnique(languages));
    }
}
